feat: add support to generate only the latest versions

dandhlee opened the pull request:
Adding an environment variable `BUILD_LATEST` to allow refreshing only the latest doc. For `FORCE_GENERATE_ALL` or `LANGUAGE` builds, if `BUILD_LATEST` is set to true then finds all relevant blobs, finds unique packages and finds its latest versions. There's already algorithm to find latest version in the xref, so tweaked it a bit to find latest semver version for more general usage.

Not sure how to test this with given test structure, will update test case soon.

Fixes #112

dandhlee commented:
Adding do not merge until unit test is updated.

tbpg commented on README.md:
```diff
@@ -125,6 +125,9 @@ See `.trampolinerc` for the canonical list of relevant environment variables.
 * `FORCE_GENERATE_ALL`: Set to `true` to regenerate all docs.
+* `BLOB_TO_DELETE`: Blob to delete from storage. Include full bucket and object name. For
+  example: `gs://my-bucket/docfx-python-test-tarball.tar.gz`
+* `BUILD_LATEST`: Set to `true` to only regenerate latest versions for each package.
```
These options feel a bit funny to me:

* `FORCE_GENERATE_ALL` generates everything. Set `BUILD_LATEST` to only do latest.
* `LANGUAGE` generates everything for that language. Set `BUILD_LATEST` to only do latest.

Maybe `LANGUAGE` should the the "optional" item, instead of `BUILD_LATEST`? Then, we could have `FORCE_GENERATE_ALL` and `FORCE_GENERATE_LATEST`, optionally set `LANGUAGE` to restrict to just that language.

Or maybe `BUILD_LATEST` should be named `ONLY_LATEST`? We're still left with the problem it's easy to forget the behavior of `LANGUAGE` -- does that do all for that language?

I guess the question is, what is more intuitive/easier to understand?

1. `FORCE_GENERATE_LATEST=true` `LANGUAGE=foo`
2. `LANGUAGE=foo ONLY_LATEST=true`

I think I like option 1, but I could be convinced.

dandhlee replied:
How about if we separate this into a separate Kokoro job, that only builds latest? So instead of `generate-prod`, we'll have `generate-prod-latest` where it'll generate `FORCE_GENERATE_LATEST` by default, and an optional `LANGUAGE=lang` that users can supply if they only want to build latest doc for a language. 

The motivation behind a separate Kokoro job would be to keep the environment variable in Kokoro only to one. The new job can be focused on only building the latest docs, while the existing Kokoro job doesn't have to be modified at all. It has a downside of having to keep track of another Kokoro job for developers.

If we keep this to one Kokoro job, I think both options are equally weighted in my preference. 
```
FORCE_GENERATE_ALL=true, LANGUAGE=lang
FORCE_GENERATE_ALL=true
FORCE_GENERATE_LATEST=true, LANGUAGE=lang
FORCE_GENERATE_LATEST=true
```
or
```
LANGUAGE=lang
LANGUAGE=lang, ONLY_LATEST=true
FORCE_GENERATE_ALL=true
FORCE_GENERATE_ALL=true, ONLY_LATEST=true
```
* Option 1 makes `FORCE_GENERATE_ALL` builds easier, slightly more tacky when you only want to build for single language.
* Option 2 makes `LANGUAGE` builds easier, slightly more tacky when you want to regenerate every latest content.

tbpg replied:
Let's keep a single job for now. If we do parallel builds, we'll make new jobs anyway.

Looking at them now, I think I prefer the first one you gave?

dandhlee replied:
Got it.

tbpg commented on generate.sh:
```diff
@@ -31,9 +31,9 @@ if [ -z "$SOURCE_BUCKET" ]; then
 
 if [ "$FORCE_GENERATE_ALL" == "true" ]; then
-    python3 docpipeline/__main__.py build-all-docs $SOURCE_BUCKET
+    python3 docpipeline/__main__.py build-all-docs $SOURCE_BUCKET $BUILD_LATEST
```
I think I'd prefer to put the logic into the Bash rather than in the Python. What if `BUILD_LATEST` isn't set? What happens with the argument? What if we add an argument after that?

If we go the `LANGUAGE`-as-the-option route from above, perhaps `--language` should be an argument?

dandhlee replied:
Will update this logic to be reflected on the bash script rather than on Python as much as possible, depending on which design we move forward with.

dandhlee commented:
Ready for review!

* Updated to use option 1 as discussed above; using `FORCE_GENERATE_LATEST` with `LANGUAGE` environment variable as optional variable
* Separated `docpipeline/__main__.py` into `build_all/latest_docs` and `build_{latest_}language_docs` to handle the environment variable logic in bash, and simply call for the corresponding generation functions in `docpipeline/generate.py` with expected behavior in the optional parameters.
* Added test case for generating the latest docs. Added unit tests by adding another version package with newer version metadata.
* Updated README

tbpg commented on tests/test_generate.py:
```diff
@@ -277,6 +293,41 @@ def test_generate(yaml_dir, tmpdir):
+    latest_html_blob_name = "python-doc-pipeline-test-2.1.2.tar.gz"
+
+    upload_yaml(yaml_dir, test_bucket)
+    generate.build_all_docs(test_bucket, storage_client, only_latest)
```
Consider `only_latest=True` to avoid the redirection of defining `only_latest` as a separate variable. When you're reading just this line (or the one below), you don't know the value of `only_latest`.

tbpg commented on tests/test_generate.py:
```diff
@@ -277,6 +293,41 @@ def test_generate(yaml_dir, tmpdir):
+    t2_latest = latest_html_blob.updated
+    assert t1_latest != t2_latest
+
+    swap_file(yaml_dir, yaml_dir / "docs.metadata", yaml_dir / new_metadata)
```
Consider swapping, uploading, and swapping back in consecutive lines to make it easy to follow.

tbpg reviewed (changes requested):
Nice!

dandhlee commented on docpipeline/generate.py:
```diff
@@ -317,6 +307,58 @@ def version_sort(v):
+    for pkg in packages:
+        for language in packages[pkg]:
+            prefix = f"{DOCFX_PREFIX}{language}-{pkg}-"
+            blobs = bucket.list_blobs(prefix=prefix)
```
Should have addressed the other comment first, changed to re-using the list of blobs.

dandhlee commented:
Updated unit test as well, please take a look again!

tbpg commented on docpipeline/generate.py:
```diff
@@ -317,6 +307,58 @@ def version_sort(v):
+# Returns a list of blobs of their latest versions.
+def find_latest_blobs(bucket, blobs):
+    latest_blobs = []
+    packages = {}
```
I'd prefer the other way (tarballs are lang-pkg-version; we think about languages have multiple packages, not packages having multiple languages). But, it's fine to leave as-is.

dandhlee replied:
Changed it around while fixing the other comment.

tbpg commented on tests/test_generate.py:
```diff
@@ -277,6 +293,44 @@ def test_generate(yaml_dir, tmpdir):
+    new_metadata = "docs.metadata.newer"
+    # Swap to newer metadata to upload newer version of tarball.
+    swap_file(yaml_dir, yaml_dir / "docs.metadata", yaml_dir / new_metadata)
+    latest_html_blob_name = "python-doc-pipeline-test-2.1.2.tar.gz"
```
Nit: any particular reason to have this in between the swap-upload-swap?

dandhlee replied:
Nay, I'll move it up.
